src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Telemetry/ITelemetryReporter.cs

```diff
@@ -8,11 +8,19 @@ namespace Microsoft.AspNetCore.Razor.Telemetry;
 
 internal interface ITelemetryReporter
 {
-    void InitializeSession(string telemetryLevel, string? sessionId, bool isDefaultSession);
-    IDisposable BeginBlock(string name, Severity severity);
-    IDisposable BeginBlock(string name, Severity severity, ImmutableDictionary<string, object?> values);
-    IDisposable TrackLspRequest(string lspMethodName, string lspServerName, Guid correlationId);
+    TelemetryScope BeginBlock(string name, Severity severity);
+    TelemetryScope BeginBlock(string name, Severity severity, Property property);
+    TelemetryScope BeginBlock(string name, Severity severity, Property property1, Property property2);
+    TelemetryScope BeginBlock(string name, Severity severity, Property property1, Property property2, Property property3);
+    TelemetryScope BeginBlock(string name, Severity severity, ImmutableArray<Property> properties);
```

Comment:
Why did you choose an `ImmutableArray<T>` over say `IEnumerable<T> / IReadOnlyList<T>`? In all the implementations in this PR it appears that the properties are not stored but just quickly enumerated. Saw at least one place where we are effectively building up an `ImmutableArray<T>` only to just enumerate it and discard. 

Is there an implementation that I'm missing where the `ImmutableArray<T>` is necessary?

Reply:
Enforcing `ImmutableArray<T>` allows callers to use `PooledArrayBuilder<T>` which can be much more efficient and uses stack space for storage for small arrays. With telemetry, I'm trying to avoid callers from using whatever data structure is at hand, such as `List<T>`, which would box when enumerated.

Reply:
I'd say `IEnumerable<Property>` would work great here. I don't think immutable is necessary

Reply:
Couldn't we still use a `PooledArrayBuilder<T>`? Aren't those `IEnumerable<T>`?

Reply:
We can't use `PooledArrayBuilder<T>` for a method like this. `PooledArrayBuilder<T>` has ref semantics and is intended to be short-lived. This API wants to be passed a data structure that it can hold onto until the end of its lifetime. Can't use `PooledArrayBuilder<T>` for that.

I worry that `IEnumerable<Property>` will put folks in a pit a failure. It would be too easy to pass a collection, such as List<Property> which would box the enumerator. I actually think the best thing is to remove the `ImmutableArray<Property>` APIs and just add a params API as the catch all.

Reply:
This should be improved in the latest changes.